### PR TITLE
Add Real Space

### PR DESCRIPTION
**Pull request by jhale**
Builds on previous work of @mscroggs.

Currently brings in tensor related branches from @garth-wells so would need to wait before merging.

### CODE REVIEW

File: cpp/demo/real/main.cpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#include "real.h"
```
Comment (jhale) [inline cpp/demo/real/main.cpp:1]: I will remove this, was useful for debugging and looking at generated code.

File: cpp/demo/real/real.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from ufl import FiniteElement, triangle, TrialFunction, TestFunction, dx, VectorElement, inner
```
Comment (jhale) [inline cpp/demo/real/real.py:1]: To remove.

File: cpp/dolfinx/fem/CoordinateElement.cpp [hunk 3]
```diff
@@ -50,20 +51,23 @@ xt::xtensor<double, 4>
...
-  return _element->tabulate(n, X, {X.shape(0), X.shape(1)});
+  auto [tab, shape] = _element->tabulate(n, X, {X.shape(0), X.shape(1)});
+  return xt::adapt(tab, shape);
 }
 //--------------------------------------------------------------------------------
 void CoordinateElement::tabulate(int n, const xt::xtensor<double, 2>& X,
                                  xt::xtensor<double, 4>& basis) const
 {
-  _element->tabulate(n, X, basis);
+  assert(_element);
```
Comment (jhale) [inline cpp/dolfinx/fem/CoordinateElement.cpp:61]: From Garth's branch.

File: cpp/demo/real/main.cpp (hunk 1 above)
Comment (jhale) [inline cpp/demo/real/main.cpp:1]: To remove.

**Comment by michalhabera** [comment]
What is the parallel ownership of real element dofs?

For performance reasons we could pack coefficients of Real element in a similar way to how Constants are packed - outside of the main assembly loop, just once per domain. Legacy approach to Real element has many bottlenecks, I thought we opted for redesign.

**Comment by tbrandvik** [comment]
I have some old Dolfin-based code that makes use of real spaces that I am trying to port to DolfinX. I was wondering if the capability in this branch is likely to be merged with the main branch in the near future or if I am better off using this branch directly? Thanks!

**Comment by jhale** [comment]
Perhaps not @tbrandvik, because I agree with @michalhabera that there are better ways of doing this. I have a research problem that required this implementation and I had to get it done quickly (it worked for my purposes).

I would add that this branch does not currently work in parallel, and the inner product of a real function and a finite element function in a form is also not giving the right output. So be careful!

**Comment by tbrandvik** [comment]
Thanks @jhale - that is useful information. Hopefully support for real spaces become part of the main branch at some point!

**Comment by GSadashivaiah27** [comment]
Hi,

As I understand real space is still not available within DolfinX? or Is there a way to use it.

**Comment by thisch** [comment]
I'm also interested in this feature. BTW Are there plans when the tensor related branches by @garth-wells will be merged?

**Comment by ramrebol** [comment]
I'm also interested in this feature. Do you think it will be incorporated?

**Comment by garth-wells** [comment]
It will be added at some point (hopefully soon). I still needs some careful thought on how to do it while maintaining decent performance.

**Comment by ramrebol** [comment]
thanks @garth-wells for answer. May you suggest me how to impose the integral of the pressure equal to zero in a Stokes problem? (Typical stationary Stokes problem with Dirichlet boundary condition only).

**Comment by jkrokowski** [comment]
Any status updates on this issue? I have an application in which I would like to use real type spaces.

**Comment by garth-wells** [comment]
We need to come back to this with a focused effort. Closing for now.

**Pull request closed without merging**